File: solutions/python/delete-node-with-given-key.py
def delete_node(head, key):
  prev = None
  current = head
  
  while (current != None):
    if current.data == key:
      if current == head:
        head = head.next
        current = head
      else:  
        prev.next = current.next
        current = current.next
    else:
      prev = current
      current = current.next

  # key not found in list
  if current == None:
    return head
  return head

list_head = create_random_list(10)
print("Original:")
display(list_head)

lst = to_list(list_head)

print("\nDeleting " + str(lst[5]), end="")
list_head = delete_node(list_head, lst[5])
print("\nAfter Delete LinkedListNode:")
display(list_head)

print("\nDeleting (Non-Existing) " + str(101), end="")
list_head = delete_node(list_head, 101)
print("\nAfter Delete LinkedListNode:")
display(list_head)

print("\nDeleting 1st node:" + str(lst[0]), end="")
list_head = delete_node(list_head, lst[0])
print("\nAfter Delete 1st LinkedListNode:")
display(list_head)

lst = to_list(list_head)
print("\nDeleting last node:" + str(lst[-1]), end="")
list_head = delete_node(list_head, lst[-1])
print("\nAfter Delete last LinkedListNode:")
display(list_head)

File: solutions/javascript/delete-node-with-given-key.js
let delete_node = function(head, key) {
    let prev = null;
    let current = head;

    while (current) {
        if (current.data === key) {
            if (current === head) {
                head = head.next;
                current = head;
            } else {
                prev.next = current.next;
                current = current.next;
            }
        } else {
            prev = current;
            current = current.next;
        }
    }
    
    if (!current) {
        return head;
    }
    return head;
};

console.log("");
console.log("");
console.log("+++++++++++++++++++++++++++++++++++++++");
console.log("Delete Key from LinkList");
console.log("---------------------------------------");

let head_before_delete = create_linked_list([0, 1, 2, 3, 4, 5, 10, 9]);

let temp_head = head_before_delete;
console.log("before delete key");
while (temp_head) {
    console.log(temp_head.data);
    temp_head = temp_head.next;
}

head_before_delete = delete_node(head_before_delete, 5);
temp_head = head_before_delete;
console.log("after delete key number 5");
while (temp_head) {
    console.log(temp_head.data);
    temp_head = temp_head.next;
}
head_before_delete = delete_node(head_before_delete, 10);
temp_head = head_before_delete;
console.log("after delete key number 10");
while (temp_head) {
    console.log(temp_head.data);
    temp_head = temp_head.next;
}
head_before_delete = delete_node(head_before_delete, 0);
temp_head = head_before_delete;
console.log("after delete key number 0");
while (temp_head) {
    console.log(temp_head.data);
    temp_head = temp_head.next;
}
